### PR TITLE
Create read session in estimateStatistics and use metadata from reponse

### PR DESCRIPTION
**Pull request by nileyadav**

### CODE REVIEW

File: spark-bigquery-dsv2/spark-bigquery-dsv2-common/src/main/java/com/google/cloud/spark/bigquery/v2/context/BigQueryDataSourceReaderContext.java [hunk 1]
```diff
@@ -311,9 +324,28 @@ public void pruneColumns(StructType requiredSchema) {
   }
 
   public StatisticsContext estimateStatistics() {
-    return table.getDefinition().getType() == TableDefinition.Type.TABLE
-        ? new StandardTableStatisticsContext(table.getDefinition())
-        : UNKNOWN_STATISTICS;
+    createReadSession();
+    if (table.getDefinition().getType() == TableDefinition.Type.TABLE) {
```
Comment (davidrabinowitz) [inline spark-bigquery-dsv2/spark-bigquery-dsv2-common/src/main/java/com/google/cloud/spark/bigquery/v2/context/BigQueryDataSourceReaderContext.java:328]: No need to create StandardTableDefinition for this. Statistics is a simple interface with two methods - you can create an anonymous implementation that returns the data from the readsession (or a simple class with two members

**Review by davidrabinowitz — CHANGES_REQUESTED** [review]

File: spark-bigquery-dsv2/spark-bigquery-dsv2-common/src/main/java/com/google/cloud/spark/bigquery/v2/context/BigQueryDataSourceReaderContext.java (hunk 1 above)
Comment (nileyadav) [reply spark-bigquery-dsv2/spark-bigquery-dsv2-common/src/main/java/com/google/cloud/spark/bigquery/v2/context/BigQueryDataSourceReaderContext.java:321]: Created anonymous implementation of StatisticsContext

File: spark-bigquery-dsv2/spark-bigquery-dsv2-common/src/main/java/com/google/cloud/spark/bigquery/v2/context/BigQueryDataSourceReaderContext.java [hunk 2]
```diff
@@ -269,6 +253,35 @@ private ReadRowsResponseToInternalRowIteratorConverter createConverter(
...
+  // We need to create read session in or before estimateStatistics,
+  // so that metadata from read session response can be saved as table statistics and
+  // same is returned to Spark.
+  // "estimateStatistics" is called before "planBatchInputPartitionContexts" or
+  // "planInputPartitionContexts" in Java Spark 3.1 connector.
+  // But Java Spark 2.4 connector does not have same calling order.
+  // As this code is common for Spark 3.1 and 2.4, making this method "synchronized" to make sure
+  // readSessionResponse is created only
+  // once.
+  private synchronized void createReadSession() {
```
[the remark this answers is not in the document]
Comment (nileyadav) [reply spark-bigquery-dsv2/spark-bigquery-dsv2-common/src/main/java/com/google/cloud/spark/bigquery/v2/context/BigQueryDataSourceReaderContext.java]: It is simpler! Thank you for the suggestion.

**Review by davidrabinowitz — APPROVED** [review]

**Pull request merged**